src/main/java/com/hivemq/extensions/handler/ConnectInboundInterceptorHandler.java

```diff
@@ -102,170 +90,184 @@ public ConnectInboundInterceptorHandler(
-                        channel, logMessage, logMessage, Mqtt5ConnAckReasonCode.UNSPECIFIED_ERROR,
+        public void pluginPost(final @NotNull ConnectInboundOutputImpl output) {
+            if (output.isPrevent()) {
+                finishInterceptor();
+            } else if (output.isTimedOut() && (output.getTimeoutFallback() == TimeoutFallback.FAILURE)) {
+                output.prevent(
+                        "Connect with client ID " + getIdentifier() + " failed because of an interceptor timeout",
                         "Extension interceptor timeout");
-                interceptorFuture.set(null);
-                return;
-            }
-
-            if (pluginOutput.getConnectPacket().isModified()) {
-                input.updateConnect(pluginOutput.getConnectPacket(), configurationService);
+                finishInterceptor();
+            } else {
+                if (output.getConnectPacket().isModified()) {
+                    inputHolder.set(inputHolder.get().update(output));
+                }
+                if (!finishInterceptor()) {
+                    outputHolder.set(output.update(inputHolder.get()));
+                }
             }
+        }
 
+        public boolean finishInterceptor() {
             if (counter.incrementAndGet() == interceptorCount) {
-                interceptorFuture.set(null);
+                ctx.executor().execute(this);
+                return true;
             }
+            return false;
         }
 
-        public void increment() {
-            //we must set the future when no more interceptors are registered
-            if (counter.incrementAndGet() == interceptorCount) {
-                interceptorFuture.set(null);
+        @Override
+        public void run() {
+            final ConnectInboundOutputImpl output = outputHolder.get();
+            if (output.isPrevent()) {
+                final String logMessage = output.getLogMessage();
+                final String reasonString = output.getReasonString();
+                connacker.connackError(
+                        ctx.channel(), logMessage, logMessage, Mqtt5ConnAckReasonCode.UNSPECIFIED_ERROR, reasonString);
+            } else {
+                final CONNECT connect = CONNECT.from(inputHolder.get().getConnectPacket(), hivemqId.get());
+                ctx.channel().attr(ChannelAttributes.CLIENT_ID).set(connect.getClientIdentifier());
+                ctx.channel()
+                        .attr(ChannelAttributes.PLUGIN_CLIENT_INFORMATION)
+                        .set(new ClientInformationImpl(connect.getClientIdentifier()));
+                ctx.channel().attr(ChannelAttributes.CLEAN_START).set(connect.isCleanStart());
+                ctx.channel().attr(ChannelAttributes.CONNECT_KEEP_ALIVE).set(connect.getKeepAlive());
+                ctx.channel().attr(ChannelAttributes.AUTH_USERNAME).set(connect.getUsername());
+                ctx.channel().attr(ChannelAttributes.AUTH_PASSWORD).set(connect.getPassword());
```

Comment:
WILL_PRESENT could also change

Reply:
@LBrandl I cannot find a WILL_PRESENT channel attribute.
Also this is the same code as before, just moved to another place